src/login/logind-session-device.c

```diff
@@ -103,6 +103,82 @@ static void sd_eviocrevoke(int fd) {
         }
 }
 
+#if BPF_FRAMEWORK
+#include "bpf-dlopen.h"
+#include "bpf/logind-revoke-skel.h"
+
+static struct logind_revoke_bpf *logind_revoke_bpf_free(struct logind_revoke_bpf *obj) {
+        /* logind_revoke_bpf__destroy handles object == NULL case */
+        (void) logind_revoke_bpf__destroy(obj);
+
+        return NULL;
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct logind_revoke_bpf *, logind_revoke_bpf_free);
+
+static int hidraw_bpf_attach(struct logind_revoke_bpf **ret_obj) {
+        _cleanup_(logind_revoke_bpf_freep) struct logind_revoke_bpf *obj = NULL;
+        int r;
+
+        assert(ret_obj);
+
+        r = dlopen_bpf();
+        if (r < 0) {
+                log_info_errno(r, "Failed to open libbpf, revoke for hidraw is not supported: %m");
+                return -ENOTSUP;
+        }
+
+        obj = logind_revoke_bpf__open();
+        if (!obj)
+                return log_error_errno(errno, "Failed to open BPF object: %m");
+
+        r = logind_revoke_bpf__load(obj);
+        if (r < 0)
+                return log_error_errno(-r, "Failed to load BPF object");
+
+        *ret_obj = TAKE_PTR(obj);
+
+        return 0;
+}
+
+struct hidraw_revoke_syscall_args {
+        int fd;
+};
+
+static void sd_hidraw_revoke(int fd) {
+        struct hidraw_revoke_syscall_args args = { .fd = fd };
+        _cleanup_(logind_revoke_bpf_freep) struct logind_revoke_bpf *obj = NULL;
+        int r;
+
+        DECLARE_LIBBPF_OPTS(bpf_test_run_opts, tattrs,
+                        .ctx_in = &args,
+                        .ctx_size_in = sizeof(args),
+                        );
+
+        assert(fd >= 0);
+
+        r = hidraw_bpf_attach(&obj);
+        if (r == 0) {
+                int bpf_fd = sym_bpf_program__fd(obj->progs.hidraw_revoke);
+                r = sym_bpf_prog_test_run_opts(bpf_fd, &tattrs);
+                if (r < 0)
+                        log_warning_errno(-r, "Revoking hidraw failed: %m");
```

Comment:
no need to add `-`. our log functions take abs() of the param anyway, hence you can pass either positive or negative doesn't atter